### PR TITLE
libargon2: fixed cross-compilation

### PR DESCRIPTION
**Pull request by vikanezrimaya**
###### Motivation for this change
Makefile had a hardcoded unprefixed ar. I wrote a patch and added an optional make flag to override it in case we're cross-compiling.

Unfortunately, this causes a rebuild of native packages.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Radvendii @olynch

### CODE REVIEW

**Comment by aanderse** [comment]
Looks like this patch should be an upstream fix. Can you open a ticket upstream please?

**Comment by kisik21** [comment]
@aanderse Done! P-H-C/phc-winner-argon2#277

**Review by aanderse — CHANGES_REQUESTED** [review]
That is great you could submit the fix upstream! Thanks!

File: pkgs/development/libraries/libargon2/default.nix [hunk 1]
```diff
@@ -11,6 +11,12 @@ stdenv.mkDerivation rec {
     sha256 = "0p4ry9dn0mi9js0byijxdyiwx74p1nr8zj7wjpd1fjgqva4sk23i";
   };
 
+  # Fix cross-compilation
+  patchPhase = ''
+    patch Makefile ${./ar.patch}
+  '';
+  makeFlags = [] ++ stdenv.lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) ["AR=${stdenv.hostPlatform.config}-ar"];
```
[the remark this answers is not in the document]
Comment (kisik21) [reply pkgs/development/libraries/libargon2/default.nix]: @aanderse The first way seems more logical for me, as (I suppose) it preserves makeFlags as a list. Patch incoming soon!

File: pkgs/development/libraries/libargon2/default.nix [hunk 2]
```diff
@@ -11,6 +11,12 @@ stdenv.mkDerivation rec {
     sha256 = "0p4ry9dn0mi9js0byijxdyiwx74p1nr8zj7wjpd1fjgqva4sk23i";
   };
 
+  # Fix cross-compilation
+  patchPhase = ''
```
[the remark this answers is not in the document]
Comment (kisik21) [reply pkgs/development/libraries/libargon2/default.nix]: Ok, I'll take note of it.

**Comment by kisik21** [comment]
Accidentally forgot to commit stuff when rewriting my commit. Sorry! :3

**Comment by kisik21** [comment]
Ugh... I'm so clumsy when it comes to rewriting my pull requests. I'm terribly sorry for overloading OfBorg

**Review by aanderse — APPROVED** [review]
Looking good. Hopefully someone who knows something about cross compilation can approve. Maybe @matthewbauer?

File: pkgs/development/libraries/libargon2/default.nix [hunk 3]
```diff
@@ -11,6 +11,17 @@ stdenv.mkDerivation rec {
...
+  patches = [
+    # TODO: remove when https://github.com/P-H-C/phc-winner-argon2/pull/277 is merged + released
+    (fetchpatch {
+      url = "https://github.com/P-H-C/phc-winner-argon2/commit/cd1c1d8d204e4ec4557e358013567c097cb70562.patch";
+      sha256 = "169pyplxlbznxkacpssbj9aanh24bqfgff47ar8zgdqs31174xip";
+    })
+  ];
+
+  # Fix cross-compilation
+  makeFlags = stdenv.lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) ["AR=${stdenv.hostPlatform.config}-ar"];
```
[the remark this answers is not in the document]
Comment (kisik21) [reply pkgs/development/libraries/libargon2/default.nix]: @matthewbauer I assume that ${stdenv.cc.targetPrefix} is a no-op when natively-building?

[the remark this answers is not in the document]
Comment (kisik21) [reply pkgs/development/libraries/libargon2/default.nix]: Done!

**Comment by kisik21** [comment]
@worldofpeace Your patches were successfully assimilated :smile_cat:

**Comment by kisik21** [comment]
Oops! I've force-pushed it to my commit >.< One second, I'll amend the commit message.

**Comment by kisik21** [comment]
Noted the pkg-config fix in the commit message and attributed the patch.

**Review by worldofpeace — APPROVED** [review]
Checked that it cross compiles for: `aarch64-multiplatform` `raspberryPi`.

File: pkgs/development/libraries/libargon2/default.nix [hunk 4]
```diff
@@ -11,18 +11,22 @@ stdenv.mkDerivation rec {
...
+    })
+  ];
 
-    make install PREFIX=$out LIBRARY_REL=lib
-    ln -s $out/lib/libargon2.so $out/lib/libargon2.so.0
-    runHook postInstall
-  '';
+  # Fix cross-compilation
+  makeFlags = [
+    "AR=${stdenv.cc.targetPrefix}ar"
```
[the remark this answers is not in the document]
Comment (kisik21) [reply pkgs/development/libraries/libargon2/default.nix]: @worldofpeace done

**Pull request merged**